oemof_b3/tools/testing_pipeline.py

```diff
@@ -163,83 +181,169 @@ def get_abs_path_list(output_rule_list):
 
     Outputs
     -------
-    absolute_path_list : str
-        Absolute file path
+    absolute_path_list : list
+         Absolute file path in list
 
     """
-    # Loop over each rule which is tested in the snakemake pipeline
-    for sublist in output_rule_list:
-        # Get absolute path
-        absolute_path_list = [os.path.join(os.getcwd(), entry) for entry in sublist]
+    # Get absolute path of rule
+    absolute_path_list = [os.path.abspath(entry) for entry in output_rule_list]
 
     return absolute_path_list
 
 
+def file_name_extension(raw_file_path):
+    """
+    This function rearranges the current absolute file path
+    with the new extension suffix '_original'.
+
+    Inputs
+    -------
+    raw_file_path : str
+        Absolute path of rule
+
+    Outputs
+    -------
+    renamed_path : str
+
+    """
+    # Get file extension
+    file_extension = raw_file_path[raw_file_path.rfind(".") + 1 :]
+    # Rename existing user data
+    renamed_file = rename_path(
+        raw_file_path, "." + file_extension, "_original." + file_extension
+    )
+
+    return renamed_file
+
+
+def rule_test(sublist):
+    """
+    This function runs the rule from the output rule sublist.
+
+    Inputs
+    -------
+    sublist : str
+        Path of rule
+
+    Outputs
+    -------
+    None
+
+    """
+    # Run the snakemake rule in this loop
+    output = snakemake.snakemake(
+        targets=sublist,
+        snakefile="Snakefile",
+    )
+
+    # Check if snakemake rule exited without error (true)
+    assert output, f"Snakemake rule failed for targets: {sublist}"
+
+    # Log the success
+    logging.info(f"Snakemake rule executed successfully for targets: {sublist}")
+
+
+def clean_file(sublist, delete_switch, renamed_path):
+    """
+    This function removes test data files and reverts renamed files.
+
+    Inputs
+    -------
+    sublist : list of str
+        List of target paths of rules
+    delete_switch : bool
+        If True, delete the data created during the test run.
+        If False, do not delete the data.
+    renamed_path : list of str
+        List of renamed target paths
+
+    Outputs
+    -------
+     None
+
+    """
+    # Remove the file created for this test
+    for raw_file_path in sublist:
+        if os.path.exists(raw_file_path):
+            if delete_switch or renamed_path:
+                remove_test_data(raw_file_path)
+
+    # If file had to be renamed revert the changes
+    for renamed_file in renamed_path:
+        if os.path.isfile(renamed_file):
+            file_extension = renamed_file[renamed_file.rfind(".") + 1 :]
+            rename_path(
+                renamed_file,
+                "_original." + file_extension,
+                "." + file_extension,
+            )
+
+
 def pipeline_file_output_test(delete_switch, output_rule_list):
+    """
+    This function tests the Snakemake pipeline for a list of output rules
+    and reverts all changes made in the target directory.
```

Comment:
So far only for files and not for dirs. But when we are going to reuse it for dir tests as well in test_pipeline_results, we should rename the function instead.

Reply:
Thank you for pointing this out! I will change the description accordingly.

I left the function `pipeline_file_output_test` as is, therefore the tests for the files stays unaffected. In the function `pipeline_folder_output_test` I adjusted the code to process both dirs and files. This needs to be tested for `test_pipeline_resources_tables.py` and `test_pipeline_raw_data.py` .

Reply:
The function has been adjusted and renamed with this commit [ca75bbd](https://github.com/rl-institut/oemof-B3/pull/357/commits/ca75bbdc446f1cae4c97cdd0d985c2bbb8ee9cc5).
The function has been tested for `test_pipeline_resources_tables.py`, `test_pipeline_raw_data.py`, and `test_pipeline_results.py`.